Loyalty extension example overhaul

js-goupil opened the pull request:
### Background
This changes the loyalty extension example to leverage Direct API + metafields instead of a Remix server + database.

### Solution
Leverages Direct API. Still needs to add points at the end of a transaction, but will figure this out tomorrow.

### 🎩
- You can try to implement this extension and try it out locally!

### Checklist

- [ ] I have :tophat:'d these changes
- [ ] I have updated relevant documentation

js-goupil committed:
Loyalty extension example overhaul

vctrchu commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+    return (
+      <POSBlock>
+        <POSBlockRow>
+          <Text color="TextWarning">Unable to fetch loyaltypoints.</Text>
```
```suggestion
          <Text color="TextWarning">Unable to fetch loyalty points.</Text>
```

vctrchu commented:
> Leverages Direct API. Still needs to add points at the end of a transaction, but will figure this out tomorrow.

I will wait for this before approving.

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+          variant="captionRegular"
+          color={pointsTotal > 0 ? 'TextSuccess' : 'TextWarning'}
+        >
+          Point Balance:{pointsTotal}
```
```suggestion
          Point Balance: {pointsTotal}
```

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+    <POSBlock>
+      <POSBlockRow>
+        <Text
+          variant="captionRegular"
```
```suggestion
          variant="body"
```

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+          {availableDiscounts.map((tier, index) => (
+            <POSBlockRow key={`${tier.pointsRequired}-${index}`}>
+              <Button
+                title={`Redeem $${tier.discountValue} Discount (Use ${tier.pointsRequired} points)`}
```
```suggestion
                title={`Redeem $${tier.discountValue} discount (${tier.pointsRequired} points)`}
```

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/shopify.extension.toml:
```diff
@@ -1,6 +1,6 @@
 # The version of APIs your extension will receive. Learn more:
 # https://shopify.dev/docs/api/usage/versioning
-api_version = "2024-10"
+api_version = "unstable"
```
Made some comments re: this file in `/shopify-dev`, not sure if they're out of sync?

js-goupil replied:
Yes they were out of sync - I'll just update the one in shopify-dev. Although as I mentioned here, we will merge this PR first and once it's in unstable, the shopify-dev PR will be updated to point to unstable again, instead of my current commit (as it is currently).

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+        </POSBlockRow>
+      ) : (
+        <POSBlockRow>
+          <Text variant="headingSmall" color="TextWarning">
```
```suggestion
          <Text variant="body" color="TextWarning">
```

runmad commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+      </POSBlockRow>
+      {availableDiscounts.length > 0 ? (
+        <POSBlockRow>
+          <Text variant="headingSmall">Available Discounts:</Text>
```
```suggestion
          <Text variant="body">Available Discounts:</Text>
```

js-goupil commented:
@runmad @vctrchu Applied all your suggestions. I also added some new logic in transaction-complete to reward points when there's no discount used. The [other PR](https://github.com/Shopify/shopify-dev/pull/57798) also points to this latest commit. In that shopify-dev PR, I also added a "Next steps" that encourages the person learning, to explore building a post purchase block and tile + modal targets to enhance the experience even more.

andy-chhuon commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/CustomerDetailsBlock.tsx:
```diff
@@ -0,0 +1,112 @@
+    );
+  }
+
+  if (pointsTotal === null) {
```
I believe this condition never happens because `useFetchLoyaltyPoints` initializes pointsTotal to 0. It displays a balance of 0 when graphql errors out.

![image](https://github.com/user-attachments/assets/892f84d3-b2d0-49c2-9539-87d6f30ee328)

andy-chhuon commented on packages/ui-extensions/docs/surfaces/point-of-sale/mdxExamples/loyalty-example/TransactionCompleteObserve.ts:
```diff
@@ -0,0 +1,84 @@
+            key: 'loyalty_direct_points',
+            namespace: 'custom',
+            ownerId: `gid://shopify/Customer/${customerId}`,
+            value: `${newPointsBalance}`,
```
I think you need to add `"type": "single_line_text_field"` here.

![image](https://github.com/user-attachments/assets/f9959a80-52ff-4cb2-adbf-e4979d8d3d20)

js-goupil replied:
Added the integer type, but that works nicely

js-goupil replied:
https://shopify.dev/docs/apps/build/custom-data/metafields/list-of-data-types

andy-chhuon reviewed (changes requested):
Tophat on IOS :tophat: Small comment, but looks good!

https://github.com/user-attachments/assets/563d3359-5be8-428d-8f84-7e5ebf236eac

andy-chhuon reviewed (approved):
Looks great, nice work!

js-goupil commented:
Merging on the assmption that the 2 approvals in the shopify-dev PR account for this PR as well.
